src/bin/util/softhsm2-util-ossl.cpp

```diff
@@ -216,6 +216,86 @@ int crypto_import_key_pair
 	return result;
 }
 
+int crypto_import_certificate
+(
+	CK_SESSION_HANDLE hSession,
+	char* filePath,
+	char* label,
+	char* objID,
+	size_t objIDLen
+)
+{
+	BIO* in = NULL;
+
+	if (!(in = BIO_new_file(filePath, "rb")))
+	{
+		fprintf(stderr, "ERROR: Could open the PKCS#8 file: %s\n", filePath);
+		return 1;
+	}
+
+	X509* x509 = PEM_read_bio_X509(in, NULL, NULL, NULL);
+	BIO_free(in);
+
+	if (!x509)
+	{
+		fprintf(stderr, "ERROR: Could not read the certificate file.\n");
+		return 1;
+	}
+
+	int blobSize = i2d_X509(x509, NULL);
```

Comment:
Its advised to check if the return length (`blobSize`) is negative, in which case an error occurred (https://linux.die.net/man/3/i2d_x509).

Also for similar calls below.

Reply:
I rework the implementation, I hope the cleanup pattern is acceptable.

Reply:
Thanks for handling this so quickly. I know it is a bit pedantic but a check `< 0` would improve things a bit more. This is because the manual (https://linux.die.net/man/3/i2d_x509) seems to mention:
>a negative value if an error occurs
not just `-1.`

Reply:
I overlooked this, fixed.